src/backend/farmer/direct_node_client.rs

```diff
@@ -0,0 +1,639 @@
+                                None
+                            }
+                            Entry::Vacant(entry) => {
+                                entry.insert(acknowledgement_sender);
+
+                                // This will be sent to the farmer
+                                Some(archived_segment.segment_header)
+                            }
+                        };
+
+                    cached_archived_segment
+                        .lock()
+                        .replace(CachedArchivedSegment::Weak(Arc::downgrade(
+                            &archived_segment,
+                        )));
+
+                    maybe_archived_segment_header
+                };
+
+                Box::pin(async move { maybe_archived_segment_header })
+            });
+
+        let archived_segment_acknowledgement_senders =
+            self.archived_segment_acknowledgement_senders.clone();
+        let fut = async move {
+            let mut archived_segment_acknowledgement_senders =
+                archived_segment_acknowledgement_senders.lock();
+
+            archived_segment_acknowledgement_senders
+                .senders
+                .remove(&subscription_id);
+        };
+
+        self.subscription_executor.spawn(
+            "subspace-archived-segment-header-subscription",
+            Some("rpc"),
+            fut.boxed(),
+        );
+
+        Ok(Box::pin(stream))
+    }
+
+    async fn acknowledge_archived_segment_header(
+        &self,
+        segment_index: SegmentIndex,
+    ) -> Result<(), Error> {
+        let archived_segment_acknowledgement_senders =
+            self.archived_segment_acknowledgement_senders.clone();
+
+        let maybe_sender = {
+            let mut archived_segment_acknowledgement_senders_guard =
+                archived_segment_acknowledgement_senders.lock();
+
+            (archived_segment_acknowledgement_senders_guard.segment_index == segment_index)
+                .then(|| {
+                    let last_key = *archived_segment_acknowledgement_senders_guard
+                        .senders
+                        .keys()
+                        .next()?;
+
+                    archived_segment_acknowledgement_senders_guard
+                        .senders
+                        .remove(&last_key)
+                })
+                .flatten()
+        };
+
+        if let Some(sender) = maybe_sender {
+            if let Err(error) = sender.unbounded_send(()) {
+                if !error.is_closed() {
+                    warn!("Failed to acknowledge archived segment: {error}");
+                }
+            }
+        }
+
+        debug!(%segment_index, "Acknowledged archived segment.");
+
+        Ok(())
+    }
+
+    async fn piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, Error> {
+        let cached_segment = {
+            let cached_archived_segment = self.cached_archived_segment.lock();
```

Comment:
This lock is not async, which means this code will block async executor for the duration of segment re-creation below. Moreover, in original code lock was intentionally held for the duration of the method, while new version doesn't, meaning that during concurrent access it is possible that you'll re-create archived segment unnecessarily twice.

Instead of taking lock twice like you did here, you should have fixed clippy complain by switching to async mutex instead (from `async-lock` crate), then code would be basically the same as it was before except a few `.await` points. Also you don't rally need to replace cached segment inside of blocking task, run just `recreate_genesis_segment` inside of blocking task and everything else can live outside of it just like it did before, we only want to move very expensive segment creation operation out of the async thread.